airflow/models/taskinstance.py

```diff
@@ -450,7 +450,16 @@ def _execute_callable(context, **execute_callable_kwargs):
         else:
             xcom_value = None
         if xcom_value is not None:  # If the task returns a result, push an XCom containing it.
-            task_instance.xcom_push(key=XCOM_RETURN_KEY, value=xcom_value, session=session)
+            if task_to_execute.multiple_outputs:
+                if not isinstance(xcom_value, Mapping):
+                    raise AirflowException(
+                        "The returned value from the task is not a Mapping. The `multiple_outputs`"
+                        "parameter is set to True, so the result has to be a Mapping."
+                    )
+                for key, value in xcom_value.items():
```

Comment:
Wondering if we should limit number of outputs here?  

Not really sure if this will cause issues. But, what if one accidentally returned so much data or maybe just a very large list of values. Would that cause performance issues?

Reply:
An alternative I can think of is using `return_value_count` instead of  just `multiple_outputs`.  But, just a suggestion.

Reply:
I would argue since it's an explicit parameter, the dag author should be responsible for making sure they aren't overloading the xcom. I believe that Airflow already doesn't limit xcom sizes, but instead it's just a warning in the documentation. It could be the same for the split values.
There is also no limit in the existing implementation for the decorator tasks: https://github.com/apache/airflow/blob/8fac799a250a222bf85d484947a9b3f2065d5ba4/airflow/decorators/base.py#L244

Reply:
We do not limit the size in single output, so I think we should not do it in multiple output case either.

Reply:
I don't see an option to resolve the conversation. please mark this as resolved.